clusterloader2/pkg/prometheus/prometheus.go

```diff
@@ -99,3 +110,51 @@ func applyManifests(
 	}
 	return nil
 }
+
+func WaitForPrometheusToBeHealthy(client clientset.Interface) error {
+	klog.Info("Waiting for Prometheus stack to become healthy...")
+	return wait.Poll(
+		checkPrometheusReadyInterval,
+		checkPrometheusReadyTimeout,
+		func() (bool, error) { return isPrometheusReady(client) })
+}
+
+func isPrometheusReady(client clientset.Interface) (bool, error) {
+	raw, err := client.CoreV1().
+		Services("monitoring").
+		ProxyGet("http", "prometheus-k8s", "9090", "api/v1/targets", nil /*params*/).
+		DoRaw()
+	if err != nil {
+		// This might happen if prometheus server is temporary down, log error but don't return it.
+		klog.Warning("error while calling prometheus api: %v", err)
+		return false, nil
+	}
+
+	var response targetsResponse
+	if err := json.Unmarshal(raw, &response); err != nil {
+		// This shouldn't happen, return error.
+		return false, err
+	}
+
+	for _, t := range response.Data.ActiveTargets {
+		if t.Health != "up" {
+			klog.Infof("Target {job=%s, pod=%s} not ready...", t.Labels["job"], t.Labels["pod"])
```

Comment:
How does registering targets work?
Is it possible that we call this function immediately and there won't be any targets yet (and thus we satisfy it), and then the new (no yet up) will reigster?

Reply:
IIUC, they're created from the ServiceMonitor (CRD) objects and it's very unlikely that the prometheus server will see no targets - even though they might be created after the prometheus server deployment, it's quite unlikely that starting up prometheus server pod will outrun creating these objects. 

Nevertheless I added a condition to expect at least # nodes ready target. That should guarantee that we won't run into the situation you described.